### PR TITLE
Add new `similarity` field to `knn` clause in `_search`

### PR DESCRIPTION
**Pull request by benwtrent**
This adds a new parameter to `knn` that allows filtering nearest neighbor results that are outside a given similarity.

`num_candidates` and `k` are still required as this controls the nearest-neighbor vector search accuracy and exploration. For each shard the query will search `num_candidates` and only keep those that are within the provided `similarity` boundary, and then finally reduce to only the global top `k` as normal.

For example, when using the `l2_norm` indexed similarity value, this could be considered a `radius` post-filter on `knn`.

relates to: https://github.com/elastic/elasticsearch/issues/84929 && https://github.com/elastic/elasticsearch/pull/93574

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/search/vectors/VectorSimilarityQuery.java [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+import java.util.Objects;
+
+import static org.elasticsearch.common.Strings.format;
+
+public class VectorSimilarityQuery extends Query {
+    private final float similarity;
+    private final float docScore;
+    private final Query innerKnnQuery;
+
+    public VectorSimilarityQuery(Query innerKnnQuery, float similarity, float docScore) {
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/search/vectors/VectorSimilarityQuery.java:34]: super minor: would it make sense to limit the type of query that is provided as an argument to `AbstractKnnVectorQuery`, given that that is what we always provide to it?

**Review by javanna — APPROVED** [review]
LGTM!!!

File: server/src/main/java/org/elasticsearch/search/vectors/VectorSimilarityQuery.java [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+import org.apache.lucene.search.Scorer;
+import org.apache.lucene.search.Weight;
+import org.elasticsearch.common.lucene.search.function.MinScoreScorer;
+
+import java.io.IOException;
+import java.util.Objects;
+
+import static org.elasticsearch.common.Strings.format;
+
+public class VectorSimilarityQuery extends Query {
```
[the remark this answers is not in the document]
Comment (benwtrent) [reply server/src/main/java/org/elasticsearch/search/vectors/VectorSimilarityQuery.java:29]: Possibly, there was some pushback that even such a thing was necessary. The implementation is so simple. 

I can open a PR to see what noise is made :)

File: server/src/main/java/org/elasticsearch/search/vectors/VectorSimilarityQuery.java [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+
+    float getDocScore() {
+        return docScore;
+    }
+
+    @Override
+    public Query rewrite(IndexReader reader) throws IOException {
+        Query rewrittenInnerQuery = innerKnnQuery.rewrite(reader);
+        if (rewrittenInnerQuery instanceof MatchNoDocsQuery) {
+            return new MatchNoDocsQuery();
```
[the remark this answers is not in the document]
Comment (benwtrent) [reply server/src/main/java/org/elasticsearch/search/vectors/VectorSimilarityQuery.java:57]: <sarcasm>
But what if I want to create many short lived objects?!?! What if I thought GC was having too easy a time!!!??
</sarcasm>

Yeah 🤦 I should just return the object.

File: server/src/main/java/org/elasticsearch/search/vectors/VectorSimilarityQuery.java (hunk 1 above)
Comment (benwtrent) [reply server/src/main/java/org/elasticsearch/search/vectors/VectorSimilarityQuery.java:34]: An, `AbstractKnnVectorQuery` is package private.

Comment (javanna) [reply server/src/main/java/org/elasticsearch/search/vectors/VectorSimilarityQuery.java:42]: good point :)

**Pull request merged**